source/mir/deser/msgpack.d

```diff
@@ -7,34 +7,878 @@ Macros:
+{
+    import core.bitop : bswap;
+
+    if (data.length < length)
+    {
+        version (D_Exceptions)
+            throw IonErrorCode.unexpectedEndOfData.ionException;
+        else
+            assert(0, IonErrorCode.unexpectedEndOfData.ionErrorMsg);
+    }
+
+    assert(length == 4 || length == 8);
+
+    if (length == 4)
+    {
+        uint v = unpackMsgPackVal!uint(data);
+        serializer.putValue((() @trusted => *cast(float*)&v)());
+    }
+    else if (length == 8)
+    {
+        // manually construct the ulong
+        ulong v = unpackMsgPackVal!ulong(data);
+        serializer.putValue((() @trusted => *cast(double*)&v)());
+    }
+
+    data.advance(length);
+}
+
+@safe pure
+private static void handleElement(S)(ref S serializer, MessagePackFmt type, scope ref const(ubyte)[] data)
+{
+    import mir.bignum.integer : BigInt;
+    sw: switch (type) 
+    {
+        // fixint
+        static foreach(v; MessagePackFmt.fixint .. (1 << 7))
+        {
+            case v:
+                // XXX: FIXME - Mir Ion is buggy with unsigned
+                // types?
+                serializer.putValue(BigInt!1(v));
+                break sw;
+        }
+
+        // fixnint
+        static foreach(v; MessagePackFmt.fixnint .. 0x100)
+        {
+            case v:
+                serializer.putValue(cast(byte)v);
+                break sw;
+        }
+
+        // fixmap
+        static foreach(v; MessagePackFmt.fixmap .. MessagePackFmt.fixmap + 0x10)
+        {
+            case v:
+                readMap(serializer, data, v & 0x0F);
+                break sw;
+        }
+
+        // fixarray
+        static foreach(v; MessagePackFmt.fixarray ..  MessagePackFmt.fixarray + 0x10)
+        {
+            case v:
+                readList(serializer, data, (v - MessagePackFmt.fixarray));
+                break sw;
+        }
+        
+        // fixext
+        static foreach(v; MessagePackFmt.fixext1 .. MessagePackFmt.fixext16 + 1)
+        {
+            case v:
+                readExt(serializer, data, 1 << (v - MessagePackFmt.fixext1));
+                break sw;
+        }
+
+        // fixstr
+        static foreach(v; MessagePackFmt.fixstr .. MessagePackFmt.fixstr + 0x20)
+        {
+            case v:
+                readStr(serializer, data, (v - MessagePackFmt.fixstr));
+                break sw;
+        }
+
+
+        case MessagePackFmt.uint8:
+            serializer.putValue(BigInt!1(data[0]));
+            data.advance(1);
+            break sw;
+        
+        case MessagePackFmt.uint16:
+            serializer.putValue(BigInt!1(unpackMsgPackVal!ushort(data)));
+            data.advance(2);
+            break sw;
+
+        case MessagePackFmt.uint32:
+            serializer.putValue(BigInt!1(unpackMsgPackVal!uint(data)));
+            data.advance(4);
+            break sw;
+        
+        case MessagePackFmt.uint64:
+            serializer.putValue(BigInt!1(unpackMsgPackVal!ulong(data)));
```

Comment:
Why BigInt? You can just put unsigned numbers inside. The same for uint32 and others.

Reply:
That one was meant to solve an extremely weird bug where some small ubytes (i.e. 0, 1, 2) would not be properly serialized within the Ion value stream. If you're able to find a workaround for that (maybe casting to uint?), feel free to change that line and add it in.

Reply:
A-ha! I figured it out. It turns out that this line was faulty:
https://github.com/libmir/mir-ion/blob/44f9d9ef852d52b5bbdc9a4727dc120f19eb17f9/source/mir/deser/msgpack.d#L304-L312

For some reason, `putValue` *really* didn't like us passing `v`. Changing that to:
```d
serializer.putValue(cast(ubyte)type);
```
fixes the issue. I confirmed it further -- it looks like the `BigInt` conversions were unnecessary after all.